### ISSUE TITLE
Add more unary operators

### ISSUE THREAD

**Issue by cstjean** [JuliaLang/julia#18139]
Julia has an [impressive list](https://github.com/JuliaLang/julia/blob/c200b4cdb9620b6df369ae3c735cf3af30b6a47f/src/julia-parser.scm) of binary operators, but the unary operators are quite lonely: `(+ - ! ¬ ~ |<:| |>:| √ ∛ ∜)`, and most of them have existing meanings. Having a lot of operators makes it easier to define nice-looking DSLs. I can't think of too many mathematical unary operators, except maybe DAGGER † (can we have post-fix unary operators?) so perhaps we could grab ~random unicode characters (☼💣❖) and call them unary?

**Comment by StefanKarpinski** [comment]
I'm confused by the thinking here. Why would we want to make some random set of characters into unary operators?

**Comment by JeffBezanson** [comment]
There are a couple postfix unary operators, namely `A'` and `A.'`.

To add more, there needs to be an argument that a character has been used as an operator e.g. in some subfield, or that its only reasonable use would be as an operator, etc. It can't just be random characters.

**Comment by cstjean** [comment]
The sunshine suggestion was fanciful, but there aren't a whole lot of clean unary operators in mathematical notation. Most of them are diacritics or some form of brackets, and neither of those are suitable for Julia. Could we at least get the Hermitian operators † and ⊹?

Perhaps that's a different problem, but the way MacroTools handles variables: `@capture(ex, f_(args__) = body_)` would also be cleaner if there was a unary operator to denote "This is a variable in a pattern matching algorithm"

**Comment by JeffBezanson** [comment]
> Could we at least get the Hermitian operators † and ⊹?

Those are both currently invalid characters in julia, so yes this definitely seems possible. `⊹` has unicode category Sm, the same as most other mathematical operators, so that's an easy case to make. `†` is punctuation, but let's see what others think. If it gets a few other votes I think we can add this.

**Comment by andyferris** [comment]
`†` is frequently used in superscript position... is there some way of doing this in Unicode?

**Comment by tkelman** [comment]
I think the way is "petition the Unicode committee and hope it gets included in Unicode N+1" ?

**Comment by TotalVerb** [comment]
@cstjean What's wrong with using `$` for that?

**Comment by cstjean** [comment]
@TotalVerb For what, variables in pattern-matching? `$` is for plugging in values. MacroTools does the inverse: it extracts parts of an expression tree.

I want to build a tree with variables in it, and manipulate the tree. Right now it looks like `tree = Node(1, 2, Node(5, Var(:x)))`. I would prefer `tree = Node(1, 2, Node(5, ⨹:x))`. I can use a macro in this particular case, so it's not the end of world. If I wanted a concise binary operation though, I would have plenty of unused, exotic operators to choose from ` ⨩ ⨪ ⨫ ⨬ ⨭ ⨮ ⨹ ⨺ ⩁ ⩂ ⩅ ⩊ ⩌ ⩏ ⩐ ⩒ ⩔ ⩖ ⩗ ⩛ `

**Comment by TotalVerb** [comment]
@cstjean Correct me if I'm wrong, but two kinds of matches are distinguished here: please match this expression, which is stored in variable x from outside; and please match something here and store it in variable x which I will use inside. It seems the former is precisely "plugging in values", so `$x`, and the latter could be just `x`?

**Comment by cstjean** [comment]
This is getting us off-topic, but I mean "matching" as in logical unification. Find a substitution that will make two trees equal. I don't think `x` would work here, but I could be wrong. Feel free to email me if you want to discuss it further.

**Comment by jebej** [comment]
Is there any way to make the dagger operator parsable as a variable name? Right now it's just invalid.

**Comment by JeffBezanson** [comment]
Yes, definitely, but we would need to decide on its syntax --- i.e. does it parse as an identifier, or some kind of operator?

**Comment by jebej** [comment]
Personally, I would use dagger as an identifier (ie part of a variable name). The complex transpose already has a symbol for it so it makes little sense to just have † do the same operation.

The exact use case I have is to store the result of something like this:
```julia
a†a = a'*a 
```

**Comment by StefanKarpinski** [comment]
That seems like a nice usage well aligned with how we already allow superscript integers and such.

**Comment by stevengj** [comment]
Also, the most sensible meaning for a `†` postfix operator would be adjoint, and we already have a postfix adjoint operator.

**Comment by chakravala** [comment]
Hi, I didn't know about this issue but I found it after posting on [discourse](https://discourse.julialang.org/t/right-handed-unary-operator/6280/4)

I think it would be really great if people could define their own unary operators. Unary operators aren't so uncommon in mathematics.

Perhaps, a special character could be used to in conjunction in order to help the parser determine that it is a unary operator. For example, `` `!`` could be factorial so that one couuld do ``5`!`` and then one could define their own unary operator, each marked by `` ` ``. Another thing could be to have both left-handed and right-handed unary operators.

For example factorial would be right-handed as in ``5`!`` but if it was defined as a left-handed it could be written as ``!`5``.

Not saying that `` ` `` is the right character for this, but it's just an idea that some sort of syntax like that might help in clarifying what is a unary operator and what is not, to help enable people to define their own unary operators for whatever they need.

**Comment by stevengj** [comment]
@chakravala, the discussion of custom unary operators is very similar to the discussion of custom infix operators — see #16985.   It's a bit of a rat's nest when you get into syntax details.

**Comment by saolof** [comment]
Alternative option: add another case for the splatting syntax. Have f...(x) desugar to (args...)->f(x,args...)

This would enable syntactically lightweight currying:

```
#Basic example:
f(x,y) = #some definition
f...(x)(y) == f(x,y)

# Use in pipelining:
x |> map...(f) |> g  |> map...(g)
```

Another option could be |...> for the version of |> that works as syntax sugar. Personally I'd prefer to have at least one function application operator which is just an operator, rather than having it be special syntax. Operators work predictably.


Of course, there is the option of designing libraries so that they naturally work well with chaining by having function curry if called with too few arguments, as mentioned in #15612. This is what most languages with a function application operator do, they use currying extensively.

**Comment by jebej** [comment]
I would like to add `\dagger` to be a valid part of a variable name. Where would I do that?

**Comment by chakravala** [comment]
You need to add the character here https://github.com/JuliaLang/julia/blob/master/src/julia-parser.scm

**Comment by stevengj** [comment]
@jebej, you would modify [here](https://github.com/JuliaLang/julia/blob/2ba3fbbd4ffb03eda89c942ce9bc422e265e3c86/src/flisp/julia_extensions.c#L134-L135).

**Comment by jebej** [comment]
Should I just add it at the end of line for primes?

```julia
(wc >= 0x2032 && wc <= 0x2037) || wc == 0x2057 || wc == 0x2020 ||
```

**Comment by rkat** [comment]
FWIW I’d like to have `⁺` and `  ⃰` from Kleene algebra, but can also see the sense in keeping the number of operators to a minimum.

**Comment by StefanKarpinski** [comment]
The ship has definitely sailed on "keeping the number of operators to a minimum"—I don't think that's ever going to be the stance. I think that if the case can be made for any operator having a standard syntax and precedence that fits in with the way Julia is parsed, we're happy to support it.

**Comment by kauesena** [comment]
I think a postfix unary operator to implement `transpose` or `permutedims` is really missing. A superscript T or t is almost universally adopted as notation for transposition without conjugation in mathematical texts. However, the Unicode symbol `ᵀ` is currently allowed as part of variable names, and making it the operator would be a breaking change... (One I'd happily support.) But perhaps there is another good symbol which isn't yet allowed which could be used.

**Comment by Jollywatt** [comment]
> A superscript T or t is almost universally adopted as notation for transposition

By the way, we can currently do this:
```julia
julia> var"'ᵀ"(a) = transpose(a)

julia> [im, 1]'ᵀ
1×2 transpose(::Vector{Complex{Int64}}) with eltype Complex{Int64}:
 0+1im  1+0im
```

[Discussion](https://github.com/JuliaLang/julia/pull/38062)

**Comment by Will-Hendrix** [comment]
if we're creating a list of desired unary operators, adding the [cross matrix](https://en.wikipedia.org/wiki/Cross_product#Conversion_to_matrix_multiplication) as a postfix operator for LinearAlgebra would be nice:  
`×(v) = [0 -v[3] v[2]; v[3] 0 -v[1]; -v[2] v[1] 0]`

**Comment by stevengj** [comment]
@Will-Hendrix, see [PSA: Julia is not at that stage of development anymore](https://discourse.julialang.org/t/psa-julia-is-not-at-that-stage-of-development-anymore/44872) — any operator that is currently parsed as binary (like `×`) will not be changed to unary in 1.x (and likely never).

**Comment by ViralBShah** [comment]
Perhaps Discussions or discourse is a better place to discuss this.
